src/build/env.py

```diff
@@ -141,15 +154,23 @@ def __enter__(self):  # type: () -> IsolatedEnvironment
         if 'PATH' in os.environ:
             exe_path.append(os.environ['PATH'])
 
-        self._replace_env('PATH', os.pathsep.join(exe_path))
-        self._replace_env('PYTHONPATH', os.pathsep.join(sys_path))
-        self._replace_env('PYTHONHOME', self.path)
-        self._pop_env('PIP_REQUIRE_VIRTUALENV')
-
         self._symlink_relative(sysconfig.get_path('include'))
         self._symlink_relative(sysconfig.get_path('platinclude'))
         self._symlink_relative(sysconfig.get_config_var('LIBPL'))
 
+        self._symlink_file(sys.executable, os.path.join('bin', 'python'))
+        exe_path.insert(0, os.path.join(self.path, 'bin'))
+
+        self._executable = sys.executable
+        sys.executable = os.path.join(self.path, 'bin', 'python')
+
+        self._replace_env('PATH', os.pathsep.join(sorted(
+            set(exe_path), key=exe_path.index
+        )))  # sorted is required for Python 2 and 3.5
```

Comment:
Can we make exe_path an ordered dict instead and then we don't have to do this dance?

Reply:
We can, however I would prefer to keep it this way. Using `sorted` is only required in Python 3.5 and older, because the order isn't kept when converting to a set. It is used here as a temporary solution.

As a maintainer, I prefer to keep workarounds like this, as long as they are simple, so that we do not need to change the code in the future. When we drop Python 2 and 3.5, we can simply removed the `sorted` usage. Maybe this is another place where you and me disagree as maintainers :confused:

Reply:
Why is sorted only required for Python 2 and 3.5? `set` is unordered in all versions of Python.